### ISSUE TITLE
Allow opt-out of dagster's `SIGTERM` handler

### ISSUE THREAD

**Issue by agon-shabi** [dagster-io/dagster#25749]
## The things I understand

1. when running a step, the dagster library installs its own `SIGINT` handler; this is presumably so that it can logically 'clean up' - e.g. record any interruptions/cancellations into the event log.

2. when you're a pod running on kubernetes, and you get deleted, the container runtime will send you a `SIGTERM`.

3. the default handler for `SIGTERM` in cpython responds by immediately committing seppuku.

4. for these reasons, we have [this code](https://github.com/dagster-io/dagster/blob/a156e3296ca9e91eed8bcd6447533b01150a4f06/python_modules/dagster/dagster/_utils/interrupts.py#L18-L19)
   * i.e. we want to make sure that if our pod gets deleted, we can intercept this `SIGTERM` in order to do the logical cleanup from (1)

## The things I recently found out

1. The python `multiprocessing.Pool` implementation will typically (at least on linux) _fork_ the calling process, in order to create workers.

2. It starts a background thread that 'maintains' the pool of workers - this thread will happily create more workers than there are tasks to process, as it doesn't appear to take the task queue into account at all.
   * If the pool is initialised with the argument `maxtasksperchild=1`, the background thread will be creating new workers at the same rate that tasks are being completed (!)

3. Each such forked worker performs the following steps, in this order:
   1. Run any handlers installed via `os.register_at_fork`
      * e.g. like [this one](https://github.com/python/cpython/blob/ffb44cd08c625dc4eda7eea058e1049d1f630491/Lib/logging/__init__.py#L277) - link is to python 3.13, but is done in 3.9, 3.10, 3.11 and 3.12 too
   2. Wait for tasks to process on a shared input queue

4. When a caller is done with their `multiprocessing.Pool` instance, and they exit its context manager, the pool will:
   1. Terminate any workers that are still alive: [code](https://github.com/python/cpython/blob/ffb44cd08c625dc4eda7eea058e1049d1f630491/Lib/multiprocessing/pool.py#L716)
      * With the expectation that they will fall onto their swords the next time they get access to some clock cycles
      * This is also why it's not 'really a problem' that the background thread may create more workers than there are tasks to process - the excess workers are simply meant to die when there is no more work.
   3. Wait for them to actually be dead: [code](https://github.com/python/cpython/blob/ffb44cd08c625dc4eda7eea058e1049d1f630491/Lib/multiprocessing/pool.py#L732)

## How these come together

In our workloads, we regularly make use of `multiprocessing.Pool` to distribute work across cores where it makes sense - and every now and then we run into the issue where:

1. Step starts, dagster installs the `SIGINT` and `SIGTERM` handlers
2. Step creates a `multiprocessing.Pool` instance, which forks the current process to create many workers (inheriting these signal handlers)
4. Step is done with the `Pool` instance, and tries to `__exit__` its context manager
5. `Pool` tries to `.terminate()` the workers that it doesn't need
6. One of these workers has just been created, and - at this point - is running a handler registered with `os.register_at_fork`. It receives a `SIGTERM`, and instead of dying, raises a `DagsterExecutionInterruptedError`. This error goes nowhere, but luckily is dumped to `stderr`. The worker process finishes running the `os.register_at_fork` handler and starts its actual work - waiting for tasks on the task queue.
7. The `Pool` then tries to `.wait()` for this worker process, but the worker is blocked waiting for tasks that never come
8. We end up in a deadlock situation.

It's rare, but we run enough workloads that it's a near certainty to occur at least twice a month.

<details>
<summary>Dummy script that reproduces the deadlocking of a `multiprocessing.Pool` by overriding the default `SIGTERM` handler</summary>

```python
import os
import signal
import time
import multiprocessing

os.register_at_fork(
    # Forked worker can get be here when getting SIGTERMed.
    # If this happens, the error goes nowhere, and the forked
    # worker proceeds to wait for its next task.
    after_in_child=lambda: time.sleep(4.0),
)

def custom_signal_handler(signo, *__):
    raise RuntimeError((signo, signal.strsignal(signo)))


signal.signal(signal.SIGTERM, custom_signal_handler)


def process(item: str) -> str:
    return f"processed({item})"


with multiprocessing.Pool(processes=3) as pool:
    pool.map(process, ["item-1"])

```

might have to run in a couple of times to hit the deadlock case

</details>

## How I'm avoiding this issue as of now

1. I'm monkey-patching the code in `dagster._utils.interrupts.setup_interrupt_handlers` so that it's a no-op.
   * This means my `multiprocessing.Pool` instances can *really* clean up their excess workers.

2. I'm building my docker images with the following directive: `STOPSIGNAL SIGINT`
   * This means that when I cancel a run from the UI, or the pod gets deleted for any other reason, the container runtime sends a `SIGINT` instead of a `SIGTERM`, and the dagster code can do the cleanup it needs to.

## How I'd like to avoid the issue

I'd like to stop monkey-patching the dagster code, and would be happy to supply an environment variable that opts out of this `SIGTERM` handler installation?

## TL;DR

Our `multiprocessing.Pool` instances sometimes end up in a deadlocked state, because the dagster `SIGTERM` handler runs at the wrong time, preventing worker cleanups.

### Dagster version

1.6.0

### Deployment type

Dagster Helm chart

### Deployment details

On-prem kubernetes v1.27.6

### Additional information

@gibsondan might have the best context for this?

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.
By submitting this issue, you agree to follow Dagster's [Code of Conduct](https://github.com/dagster-io/dagster/blob/master/.github/CODE_OF_CONDUCT.md).
